<?xml version="1.0"?>
<robot name="swing_drone" xmlns:xacro="http://www.ros.org/wiki/xacro">
 
  <!-- Dummy map link as root to avoid KDL parser warning -->
  <link name="map">
  </link>

  <!-- Fixed joint connecting map to base_link -->
  <joint name="map_to_base" type="fixed">
    <parent link="map"/>
    <child link="base_link"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link
    name="base_link">
    <inertial>
      <origin
        xyz="0 0.0307 0"
        rpy="0 0 0" />
      <mass
        value="0.49" />
      <inertia
        ixx="0.0055"
        ixy="0"
        ixz="0"
        iyy="0.01067"
        iyz="0"
        izz="0.0055" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="3.14159 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/base_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="3.14159 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/base_link_new.STL" />
      </geometry>
    </collision>
  </link>

  <link 
    name="imu_link">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.001"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <box
          size="0.01 0.01 0.01" />
      </geometry>
    </visual>
  </link>

  <joint
    name="imu_joint"
    type="fixed">
    <parent
      link="base_link" />
    <child
      link="imu_link" />
    <origin
      xyz="0 0 0.01"
      rpy="0 0 0" />
    <axis
      xyz="0 0 0" />
  </joint>

  <link
    name="fr_link">
    <inertial>
      <origin
        xyz="0 0.00839 -0.16975"
        rpy="0 0 0" />
      <mass
        value="0.0713" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/arm.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.752941176470588 0.752941176470588 0.752941176470588 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/low_face_arm.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="fr_joint"
    type="fixed">
    <origin
      xyz="0 0 0.025"
      rpy="1.5708 0 2.3562" />
    <parent
      link="base_link" />
    <child
      link="fr_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="A_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.0069037"
        rpy="0 0 0" />
      <mass
        value="0.0030458" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.96863 0.015686 0.015686 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="A_prop_joint"
    type="continuous">
    <origin
      xyz="-0.075 0.032789 -0.18136"
      rpy="-1.5708 0.28363 0" />
    <parent
      link="fr_link" />
    <child
      link="A_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="B_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.0069036"
        rpy="0 0 0" />
      <mass
        value="0.0030461" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.96863 0.015686 0.015686 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="B_prop_joint"
    type="continuous">
    <origin
      xyz="0.075 0.032789 -0.18136"
      rpy="-1.5708 -0.94937 0" />
    <parent
      link="fr_link" />
    <child
      link="B_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="br_link">
    <inertial>
      <origin
        xyz="0 0.00838518216887453 -0.169743641529002"
        rpy="0 0 0" />
      <mass
        value="0.0713" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/arm.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.752941176470588 0.752941176470588 0.752941176470588 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/low_face_arm.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="br_joint"
    type="fixed">
    <origin
      xyz="0 0 0.025"
      rpy="1.5707963267949 0 0.785398163397444" />
    <parent
      link="base_link" />
    <child
      link="br_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="C_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.00690354838979317"
        rpy="0 0 0" />
      <mass
        value="0.00304576429500371" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.968627450980392 0.0156862745098039 0.0156862745098039 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="C_prop_joint"
    type="continuous">
    <origin
      xyz="-0.075 0.032789 -0.18136"
      rpy="1.5708 1.2519 3.1416" />
    <parent
      link="br_link" />
    <child
      link="C_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="D_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.0069035097240563"
        rpy="0 0 0" />
      <mass
        value="0.0030460170283791" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.968627450980392 0.0156862745098039 0.0156862745098039 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="D_prop_joint"
    type="continuous">
    <origin
      xyz="0.075 0.032789 -0.18136"
      rpy="1.5708 0.2674 -3.1416" />
    <parent
      link="br_link" />
    <child
      link="D_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="bl_link">
    <inertial>
      <origin
        xyz="0 0.00838518201458557 -0.169743641788305"
        rpy="0 0 0" />
      <mass
        value="0.0713" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/arm.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.752941176470588 0.752941176470588 0.752941176470588 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/low_face_arm.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="bl_joint"
    type="fixed">
    <origin
      xyz="0 0 0.025"
      rpy="1.5707963267949 0 -0.785398163397454" />
    <parent
      link="base_link" />
    <child
      link="bl_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="E_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.00690385853098646"
        rpy="0 0 0" />
      <mass
        value="0.00304648811584682" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.968627450980392 0.0156862745098039 0.0156862745098039 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="E_prop_joint"
    type="continuous">
    <origin
      xyz="-0.075 0.032789 -0.18136"
      rpy="1.5708 -0.4532 -3.1416" />
    <parent
      link="bl_link" />
    <child
      link="E_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="F_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.00690341554333757"
        rpy="0 0 0" />
      <mass
        value="0.00304581013536126" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.968627450980392 0.0156862745098039 0.0156862745098039 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="F_prop_joint"
    type="continuous">
    <origin
      xyz="0.075 0.032789 -0.18136"
      rpy="-1.5708 0.44248 0" />
    <parent
      link="bl_link" />
    <child
      link="F_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="fl_link">
    <inertial>
      <origin
        xyz="0 0.00838518205590654 -0.169743641875677"
        rpy="0 0 0" />
      <mass
        value="0.0713" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/arm.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.752941176470588 0.752941176470588 0.752941176470588 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/low_face_arm.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="fl_joint"
    type="fixed">
    <origin
      xyz="0 0 0.025"
      rpy="1.5707963267949 0 -2.35619449019234" />
    <parent
      link="base_link" />
    <child
      link="fl_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="G_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.00690342839194055"
        rpy="0 0 0" />
      <mass
        value="0.00304542470661797" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.968627450980392 0.0156862745098039 0.0156862745098039 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CCW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="G_prop_joint"
    type="continuous">
    <origin
      xyz="-0.075 0.032789 -0.18136"
      rpy="-1.5708 0.26204 0" />
    <parent
      link="fl_link" />
    <child
      link="G_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>
  <link
    name="H_prop_link">
    <inertial>
      <origin
        xyz="0 0 0.00690356262000323"
        rpy="0 0 0" />
      <mass
        value="0.00304597813849953" />
      <inertia
        ixx="0.00001"
        ixy="0"
        ixz="0"
        iyy="0.00001"
        iyz="0"
        izz="0.00001" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.968627450980392 0.0156862745098039 0.0156862745098039 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://swing_drone/meshes/CW_prop.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="H_prop_joint"
    type="continuous">
    <origin
      xyz="0.075 0.032789 -0.18136"
      rpy="1.5708 0.93504 -3.1416" />
    <parent
      link="fl_link" />
    <child
      link="H_prop_link" />
    <axis
      xyz="0 0 1" />
  </joint>


</robot>